kse/src/main/java/org/kse/crypto/signing/KseSignerInformation.java

```diff
@@ -0,0 +1,303 @@
+
+        if (!StringUtils.isBlank(emailAddress)) {
+            if (StringUtils.isBlank(shortName)) {
+                shortName = emailAddress;
+            } else {
+                shortName += " <" + emailAddress + ">";
+            }
+        }
+
+        if (StringUtils.isBlank(shortName)) {
+            shortName = subject.toString();
+        }
+
+        // subject DN can be empty in some cases
+        if (StringUtils.isBlank(shortName)) {
+            shortName = cert.getSerialNumber().toString();
+        }
+
+        return shortName;
+    }
+
+    /**
+     * Extracts the signature signing time, if present, from the signature's signed attributes.
+     *
+     * @return The signing time, if present, else null.
+     */
+    public Date getSigningTime() {
+        Date signingTime = null;
+        AttributeTable signedAttributes = getSignedAttributes();
+
+        if (signedAttributes != null) {
+            Attribute signingTimeAttribute = signedAttributes.get(CMSAttributes.signingTime);
+            if (signingTimeAttribute != null) {
+                Enumeration<?> element = signingTimeAttribute.getAttrValues().getObjects();
+                if (element.hasMoreElements()) {
+                    Object o = element.nextElement();
+                    try {
+                        if (o instanceof ASN1UTCTime) {
+                            signingTime = ((ASN1UTCTime) o).getAdjustedDate();
+                        } else if (o instanceof ASN1GeneralizedTime) {
+                            signingTime = ((ASN1GeneralizedTime) o).getDate();
+                        }
+                    } catch (ParseException e) {
+                        // Users are not going to know what to do about invalid ASN.1 date structures.
+                        // So ignore the exception.
+                    }
+                }
+            }
+        }
+        return signingTime;
+    }
+
+    /**
+     * Extracts the time stamp token, if present, from the signature's unsigned
+     * attributes.
+     *
+     * @return The time stamp token as ContentInfo, if present, else null.
+     */
+    public ContentInfo getTimeStamp() {
+
+        AttributeTable unsignedAttributes = getUnsignedAttributes();
+
+        if (unsignedAttributes != null) {
+            Attribute tsTokenAttribute = unsignedAttributes.get(PKCSObjectIdentifiers.id_aa_signatureTimeStampToken);
+            if (tsTokenAttribute != null) {
+                return ContentInfo.getInstance(tsTokenAttribute.getAttributeValues()[0]);
+            }
+        }
+
+        return null;
+    }
+
+    private void verify() {
+
+        status = CmsSignatureStatus.NOT_VERIFIED;
+
+        if (cert != null) {
+
+            boolean verified = false;
+
+            try {
+                // TODO JW - Should a provider be specified for the JcaSimpleSingerInfoVerifierBuilder?
+                if (verify(new JcaSimpleSignerInfoVerifierBuilder().build(cert))) {
```

Comment:
This builder can also take a provider, but verification doesn't use private keys. Do you recommend using the explicit provider (if not null) for verification, too?

Reply:
No, that would probably only cause issues. I think it should be safe here to always use the BC provider because `cert` is a BC class and `JcaSimpleSignerInfoVerifierBuilder` as well so the public key used for verification should always be one of the BC public key classes. 

The verify code for PKCS#10 CSRs (in `Pkcs10Util`) explicitly sets the BC provider but I am not sure if this was actually necessary for hardware providers to work or not. In any case, you can do it like this for consistency reasons.